jib-cli/src/main/java/com/google/cloud/tools/jib/cli/War.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2021 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.cli;
+
+import com.google.cloud.tools.jib.api.Containerizer;
+import com.google.cloud.tools.jib.api.JibContainer;
+import com.google.cloud.tools.jib.api.JibContainerBuilder;
+import com.google.cloud.tools.jib.api.LogEvent;
+import com.google.cloud.tools.jib.api.buildplan.AbsoluteUnixPath;
+import com.google.cloud.tools.jib.cli.logging.CliLogger;
+import com.google.cloud.tools.jib.cli.war.WarFiles;
+import com.google.cloud.tools.jib.plugins.common.globalconfig.GlobalConfig;
+import com.google.cloud.tools.jib.plugins.common.logging.ConsoleLogger;
+import com.google.cloud.tools.jib.plugins.common.logging.SingleThreadedExecutor;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Multimaps;
+import com.google.common.util.concurrent.Futures;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Future;
+import picocli.CommandLine;
+
+@CommandLine.Command(
+    name = "war",
+    mixinStandardHelpOptions = true,
+    showAtFileInUsageHelp = true,
+    description = "Containerize a war")
+public class War implements Callable<Integer> {
+  @CommandLine.Spec
+  @SuppressWarnings("NullAway.Init") // initialized by picocli
+  private CommandLine.Model.CommandSpec spec;
+
+  @CommandLine.Mixin
+  @VisibleForTesting
+  @SuppressWarnings("NullAway.Init") // initialized by picocli
+  CommonCliOptions commonCliOptions;
+
+  @CommandLine.Mixin
+  @VisibleForTesting
+  @SuppressWarnings("NullAway.Init") // initialized by picocli
+  CommonContainerConfigCliOptions commonContainerConfigCliOptions;
+
+  @CommandLine.Parameters(description = "The path to the war file (ex: path/to/my-war.war)")
+  @SuppressWarnings("NullAway.Init") // initialized by picocli
+  private Path warFile;
+
+  @CommandLine.Option(
+      names = "--app-root",
+      paramLabel = "<app root>",
+      description = "The app root on the container")
+  @SuppressWarnings("NullAway.Init") // initialized by picocli
+  private Path appRoot;
+
+  @Override
+  public Integer call() {
```

Comment:
Hmm... just asking, can there be a way to remove duplicate code between war and jar?

Reply:
That is a great question. This has proven to be a little tricky. The code here looks very similar to the one the jar command but the problem is that the individual methods take in different parameters (such as Jar cli options or War cli options). At the moment, the jar cli options (like `mode` or `jvm flags`) aren't accessible to the war command. The one section that I thought was common between the two was this:
```
ArtifactProcessor processor = ArtifactProcessors.fromWar(warFile, cacheDirectories, this, 
 commonContainerConfigCliOptions);
JibContainerBuilder containerBuilder =. WarFiles.toJibContainerBuilder(
              processor, commonCliOptions, commonContainerConfigCliOptions, logger);
Containerizer containerizer = Containerizers.from(commonCliOptions, logger, cacheDirectories);
// Enable registry mirrors
Multimaps.asMap(globalConfig.getRegistryMirrors()).forEach(containerizer::addRegistryMirrors);
JibContainer jibContainer = containerBuilder.containerize(containerizer);
```
I started creating a helper class for this to serve as an `ArtifactToContainerConverter` but soon realized that in order to make this work for both cases, this class would need access to both `jar` and `war` cli options.

Reply:
Gotcha.